Sources/NIOSOCKS/Channel Handlers/SOCKSClientHandler.swift

```diff
@@ -0,0 +1,118 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+/// Data was unexpectedly written or read before the SOCKS proxy
+/// connection has been fully established.
+public struct ProxyNotEstablished: Error {
+    
+}
+
+/// Connects to a SOCKS server to establish a proxied connection
+/// to a host. This handler should be inserted at the beginning of a
+/// channel's pipeline.
+public class SOCKSClientHandler: ChannelDuplexHandler {
+    
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    public typealias OutboundIn = ByteBuffer
+    public typealias OutboundOut = ByteBuffer
+    
+    public let supportedAuthenticationMethods: [AuthenticationMethod]
+    public let targetAddress: AddressType
+    public let targetPort: UInt16
+    
+    private var state: ClientStateMachine
+    private var buffered: ByteBuffer
+    
+    public init(supportedAuthenticationMethods: [AuthenticationMethod], targetAddress: AddressType, targetPort: UInt16) {
+        precondition(supportedAuthenticationMethods.count <= 255)
+        self.supportedAuthenticationMethods = supportedAuthenticationMethods
+        self.state = ClientStateMachine()
+        self.buffered = ByteBuffer()
```

Comment:
We should take a buffer from the `ChannelHandlerContext` instead of summoning one into existence here.

Reply:
Yep, done!

Reply:
We don't have a context in the initialiser, so can't do it here. Obviously I can get around this, @Lukasa do you have an opinion?

Reply:
I still think we should DTRT here. That means making this `Optional` and allocating it in `handlerAdded` instead.

Reply:
Cool, done!